core/server/utils/image-size.js

```diff
@@ -0,0 +1,193 @@
+var sizeOf = require('image-size'),
+    url = require('url'),
+    Promise = require('bluebird'),
+    got = require('got'),
+    utils = require('../utils'),
+    errors = require('../errors'),
+    config = require('../config'),
+    storage = require('../adapters/storage'),
+    _ = require('lodash'),
+    dimensions,
+    getImageSizeFromUrl,
+    getImageSizeFromFilePath;
+
+/**
+ * @description compares the imagePath with a regex that reflects our local file storage
+ * @param {String} imagePath as URL or filepath
+ * @returns {Array} if match is true or null if not
+ */
+function isLocalImage(imagePath) {
+    imagePath = utils.url.urlFor('image', {image: imagePath}, true);
+
+    return imagePath.match(new RegExp('^' + utils.url.urlJoin(utils.url.urlFor('home', true), utils.url.getSubdir(), '/', utils.url.STATIC_IMAGE_URL_PREFIX)));
+}
+
+// Supported formats of https://github.com/image-size/image-size:
+// BMP, GIF, JPEG, PNG, PSD, TIFF, WebP, SVG, ICO
+// ***
+// Takes the url of the image and an optional timeout
+// getImageSizeFromUrl returns an Object like this
+// {
+//     height: 50,
+//     url: 'http://myblog.com/images/cat.jpg',
+//     width: 50
+// };
+// if the dimensions can be fetched, and rejects with error, if not.
+// ***
+// In case we get a locally stored image, which is checked withing the `isLocalImage`
+// function we switch to read the image from the local file storage with `getImageSizeFromFilePath`.
+// In case the image is not stored locally and is missing the protocol (like //www.gravatar.com/andsoon),
+// we add the protocol and use urlFor() to get the absolute URL.
+// If the request fails or image-size is not able to read the file, we reject with error.
+
+/**
+ * @description read image dimensions from URL
+ * @param {String} imagePath as URL
+ * @returns {Promise<Object>} imageObject or error
+ */
+getImageSizeFromUrl = function getImageSizeFromUrl(imagePath) {
+    var imageObject = {},
+        requestOptions,
+        timeout = config.get('times:getImageSizeTimeoutInMS') || 10000;
+
+    imageObject.url = imagePath;
+
+    if (isLocalImage(imagePath)) {
+        // don't make a request for a locally stored image
+        return Promise.resolve(getImageSizeFromFilePath(imagePath));
+    }
+
+    // check if we got an url without any protocol
+    if (imagePath.indexOf('http') === -1) {
+        // our gravatar urls start with '//' in that case add 'http:'
+        if (imagePath.indexOf('//') === 0) {
+            // it's a gravatar url
+            imagePath = 'http:' + imagePath;
+        }
+    }
+
+    imagePath = url.parse(imagePath);
+    requestOptions = {
+        headers: {
+            'User-Agent': 'Mozilla/5.0'
+        },
+        timeout: timeout,
+        encoding: null
+    };
+
+    return got(
+        imagePath,
+        requestOptions
+    ).then(function (response) {
```

Comment:
We could create a local fn which deals with the buffer for both cases (less code, less error handling).

Reply:
👍